### PR TITLE
Feature/811 implement node state

### PR DESCRIPTION
**Pull request by ybouilla**
**PR description**
Related to issue #836 (as designed in #811)
Implements a node state facility in order to save state from one `Round` to another. It is a minimal implementation (native optimizers are not saved and it is currently only possible to load previous state. It is working only with `Strategy` `defaultstrategy`

For the next steps we have to:
- [x] add documentation
- [x] add license at the beginning of each file 
- [x] add additional unit tests / integration tests / update existing tests

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state`/`load_state` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

**Comment by ybouilla** [comment]
**NB**: there is no checking that `state_id` has been registered in the database: dear reviewer, let me know if it is ok or should be implemented

File: fedbiomed/node/node_state_manager.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    """
+    File names should contains 2 `%s`: one for round number, the second for state_id
+    Example: VALUE = "some_values_%s_%s"
+    """
+    OPTIMIZER: str = "optim_state_%s_%s"
+
+
+class NodeStateManager:
+    #FIXME: should all Manager classes be inhereting from the same ManagerBase object?
+    def __init__(self, db_path: str):
```
Comment (mvesin) [inline fedbiomed/node/node_state_manager.py:32]: missing all docstrings in this class (and in some other places)

File: fedbiomed/node/node_state_manager.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        
+        # NOTA: constructor has been designed wrt other object handling DataBase
+        self._query: Query = Query()
+        self._node_state_base_dir: str = None  # node state base directory, where all node state related files are saved
+        self.state_id: str = None
```
Comment (mvesin) [inline fedbiomed/node/node_state_manager.py:37]: state_id and previous_state_id should probably be private variables + getters

**Comment by ybouilla** [comment]
Also, I ve seen you added an exception a try/catch statement if `_load_round_state` fails, but do we really want this behaviour? In my opinion experiment should continue even if loading state fails

**Comment by mvesin** [comment]
> Hi, I don't really agree with having _load_round_state and _save_round_state being private: it makes unit test very difficult otherwise !

Thank you Yannick for your feedback.
This method is used (and meant to be used) only within the class so I think it should be private.
I dont think this is good practice to design the class to ease the unit tests.

File: fedbiomed/node/node_state_manager.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+    """
+        # NOTA: constructor has been designed wrt other object handling DataBase
+        self._query: Query = Query()
+        # node state base directory, where all node state related files are saved
+        self._node_state_base_dir: Optional[str] = None
+        self._state_id: Optional[str] = None
+        self._previous_state_id: Optional[str] = None
+        try:
+            self._db: Table = TinyDB(db_path).table(name=NODE_STATE_TABLE_NAME, cache_size=0)
```
Comment (srcansiz) [inline fedbiomed/node/node_state_manager.py:60]: Small comment on `tinydb.Table`: All other DB implementations uses DBTable that is an interface of `Table`.  There two options to make tinydb to use custom table class 

```
self._db = TinyDB(db_path)
self._table = DBTable(name=NODE_STATE_TABLE_NAME, cache_size=0)
```

or 

```
connection = TinyDB(db_path)
connection.table_class = DBTable 
self._db = connection.table(name=NODE_STATE_TABLE_NAME, cache_size=0)
```

**Review by mvesin — APPROVED** [review]

**Comment by mvesin** [comment]
Hi @ybouilla,
Thanks again for this nice work !

I implemented the researcher-side modifications discussed today + fixed corresponding unit tests.
This is done for me.
I'm holding you back the PR, if you want to review my today changes + complete some stuff.
Ping me when ready for ultimate glance & merge.

Details of points discussed today:

1. fix a pre-existing issue in `Job.__init__()`: `self._nodes` is documented as `dict` while it is a `List[str]` (list of nodes).
    + misc refactor of `Job.__init__()`: remove parameter `nodes`, as it is not used (nor useful)

2. move calls to `Nodes.update_node_states_agent()` from `Experiment` to `Job` (this should be hidden from experiment, no reason to do so) + make method private

3. simplify and refactor `NodeStateAgent` in particular:
    - don't need to pass a `FederatedDataSet`, a list of nodes (`List[str]`) is enough
    - several simplifications (strip down un-needed actions)

4. breakpoint: move save/load of node state agent stuff from `Job` to `NodeStateAgent` (restore modularity)

5. breakpoint misc: rename to `load_state_breakpoint()` / `save_state_breakpoint()` all methods related to handling of breakpoint, to avoid confusion with node side `load_state()` (for node state, for optimizer instantiation)

**Comment by ybouilla** [comment]
Hi Marc, I think we can now Merge Pull Request!

**Pull request merged**